Create 'Passage Augmenter' node and its first module 'prev_next_augmenter'

bwook00 opened the pull request:
close #338

- Create passage_augmenter node

- Create prev_next_augmenter
https://docs.llamaindex.ai/en/stable/examples/node_postprocessor/PrevNextPostprocessorDemo/

Plus, I found some typos in passage_filter/run.py, but they were minor so I PR'd them together.

vkehfdl1 commented on autorag/utils/util.py:
```diff
@@ -310,3 +312,13 @@ def sort_by_scores(row, reverse=True):
     return list(reranked_contents), list(reranked_ids), list(reranked_scores)
+
+
+def get_cosine_similarity_scores(queries: List[str], contents_list: List[List[str]]) -> List[List[float]]:
```
We already have `calculate_cosine_similarity` function in `autorag/evalaute/metrics/util.py`
It looks like this function do the same, but got different embedding. 
Why don't you use that? You can change the function's location.

vkehfdl1 commented on autorag/nodes/passageaugmenter/prev_next_augmenter.py:
```diff
@@ -0,0 +1,52 @@
+    Add passages before and/or after the retrieved passage.
+    For more information, visit https://docs.llamaindex.ai/en/stable/examples/node_postprocessor/PrevNextPostprocessorDemo/.
+
+    :param ids_list: The list of lists of ids retrieved
+    :param all_ids_list: The list of all ids in the corpus
+    :param num_passages: The number of passages to add before and after the retrieved passage
+        Default is 1.
+    :param mode: The mode of augmentation
+        'prev': add passages before the retrieved passage
+        'next': add passages after the retrieved passage
+        'both': add passages before and after the retrieved passage
+        Default is 'next'.
+    :return: The list of lists of augmented ids
+    """
+    if mode not in ['prev', 'next', 'both']:
+        raise ValueError(f"mode must be 'prev' or 'next', but got {mode}")
+
+    augmented_ids = []
+
+    id_index_map = {id_: idx for idx, id_ in enumerate(all_ids_list)}
+
+    for sublist in ids_list:
+        new_sublist = []
+        for id_ in sublist:
+            if id_ in id_index_map:
+                idx = id_index_map[id_]
+
+                if mode == 'prev' or mode == 'both':
+                    start_idx = max(0, idx - num_passages)
+                    new_sublist.extend(all_ids_list[start_idx:idx])
+
+                new_sublist.append(id_)
+
+                if mode == 'next' or mode == 'both':
+                    end_idx = min(len(all_ids_list), idx + 1 + num_passages)
+                    new_sublist.extend(all_ids_list[idx + 1:end_idx])
+
+        augmented_ids.append(new_sublist)
+
+    return augmented_ids
```
I think this is so naive approach to implement `prev_next_augmenter` 
Since the corpus once created, how can you sure that prev row or next row is 'real' prev-next passage? 
In the`corpus.parquet` there can be numerous documents. And the two documents each can be irrelevant at all. 

I think we need to build prev-next key at the metadata for running this module.
Thus, we must include prev-next key process at the `qa creation` part before implementing prev-next-augmenter module.

If once we make a code that make prev-next metadata key in the corpus, we can make the robust prev-next module after that.

vkehfdl1 commented on autorag/nodes/passageaugmenter/base.py:
```diff
@@ -0,0 +1,53 @@
+
+        contents = fetch_contents(corpus_data, ids)
+
+        scores = get_cosine_similarity_scores(queries, contents)
```
I understand that you are sorting augmented passages again after augmentation.
But I wonder why are you using 'TF-IDF' method in here? 
I think embedding cosine sim or reranker is more reliable. 
Tell me if you can find some robust result about 'why use TF-IDF at here?'

vkehfdl1 reviewed (changes requested):
Do not rush. 
The repetitive review request on low quality code is time-consuming.
The quantity of PR or implemented module is not equal to your skill or contribution to AutoRAG.

bwook00 commented:
- Add prev_id, next_id at `add_essential_metadata` function.
   - edit make corpus code(llama index, langchain)

- remake `sample_project/data/corpus.parquet` to contain `prev_id` and `next_id` in metadata.

- add embedding model at passage_augmenter node parameter.

- make the robust prev-next module to use metadata(prev_id, next_id)

vkehfdl1 commented on autorag/data/corpus/llama_index.py:
```diff
@@ -26,11 +26,16 @@ def llama_documents_to_parquet(llama_documents: List[Document],
+            'contents': doc.text,
+            'metadata': add_essential_metadata(doc.metadata, prev_id, next_id)
+        }
+        for doc, doc_id, prev_id, next_id in zip(llama_documents, doc_ids, [None] + doc_ids[:-1], doc_ids[1:] + [None])
```
I believe that there is a Llama Document metadata that saves previous and next id.
Maybe, langchain too.
Its loader might automatically detect what is the prev or next node at TextSplitter (NodeParser at LlamaIndex), and saves to metadata to "certain key"
I don't know what is "certain key". Find that and try to use that information.

Because user can input multiple documents to one long list of Llama Documents, so it will occur that other file that has no relationship at all can detect on prev-next augmentor.

vkehfdl1 commented on autorag/nodes/passageaugmenter/prev_next_augmenter.py:
```diff
@@ -0,0 +1,57 @@
+def prev_next_augmenter(ids_list: List[List[str]],
+                        corpus_df: pd.DataFrame,
+                        num_passages: int = 1,
+                        mode: str = 'next'
```
I think it is better that default mode is 'both'.

bwook00 commented on autorag/data/corpus/llama_index.py:
```diff
@@ -60,10 +66,22 @@ def llama_text_node_to_parquet(text_nodes: List[TextNode],
+    next_node = relationships.get(NodeRelationship.NEXT, None)
+    if next_node:
+        metadata['next_id'] = next_node.node_id
+    return metadata
```
I think adding `prev_id` and `next_id` to the `validate_corpus_data` function just for the `prev_next_augmenter` is overkill.

So I created a `make_prev_next_id` function in `cast_corpus_dataset` to force the `prev_id` and `next_id` to be added (like 'last_modified_datatime').


In `llama_index_text_node_to_parquet`, if there is no `relations`, skip it, and in `cast_corpus_dataset`, add `prev_id` and `next_id`. 
(Since we're going through `cast_corpus_dataset` before using `prev_next_augmenter` anyway + decided it was unnecessarily long code. 
Instead, I added logic to not throw an error if relationships is None)

vkehfdl1 commented on autorag/data/corpus/llama_index.py:
```diff
@@ -26,11 +27,16 @@ def llama_documents_to_parquet(llama_documents: List[Document],
+            'contents': doc.text,
+            'metadata': add_essential_metadata(doc.metadata, prev_id, next_id)
+        }
+        for doc, doc_id, prev_id, next_id in zip(llama_documents, doc_ids, [None] + doc_ids[:-1], doc_ids[1:] + [None])
```
I think we cannot guarantee any prev-next relationship between each Llama Documents, right?
Just set prev_id and next_id to None, or just don't insert them since it will process at`cast_corpus_data`

vkehfdl1 commented on autorag/nodes/passageaugmenter/prev_next_augmenter.py:
```diff
@@ -0,0 +1,57 @@
+
+
+@passage_augmenter_node
+def prev_next_augmenter(ids_list: List[List[str]],
```
I think we decide to add `top_k` at passage augmentor.
Didn't you code that yet?

bwook00 replied:
Splitting into top_k should be done after `sort_by_score`, but since `sort_by_score` is done in `base.py`, I added code to take top_k as a node parameter and split it into top_k in `base.py`!

bwook00 commented:
Splitting into top_k should be done after `sort_by_score`, but since `sort_by_score` is done in `base.py`, I added code to take top_k as a node parameter and split it into top_k in `base.py`!

vkehfdl1 commented on autorag/utils/preprocess.py:
```diff
@@ -69,4 +69,20 @@ def make_datetime_metadata(x):
+        metadata.setdefault('next_id', next_doc_id)
+        return metadata
+
+    temp_prev_df = pd.DataFrame({'prev_doc_id': df['doc_id'].shift(1)})
+    temp_next_df = pd.DataFrame({'next_doc_id': df['doc_id'].shift(-1)})
```
왜 계속 이런 로직을 넣는거야?
단순히 input dataframe의 index로 어떻게 prev-next 관계를 보장하지?
내가 봤을 때는 부정확한 관계를 prev-next로 연결하는 것보다는 아예 None으로 표현하는 게 좋을 것 같은데.

corpus data 만들때부터 prev_id나 next_id가 있으면 그대로 두고, 없을 경우에는 그냥 None으로 채우는 편이 훨 좋아보이는데.
내가 이 문제에 대해서 여러번 리뷰를 달았는데 아직도 이런 로직이 있어서, 답답해서 한국어로 리뷰함. language barrier인가 해서

bwook00 replied:
`cast_corpus`의 `last_modified_time`에도 정확하진 않지만, `datetime.now()`가 박혀있으니, prev_next도 cast에서는 정확하진 않아도 무언가 확실한 값을 넣어줘야 한다고 생각했어.

처음 corpus데이터를 만들 때 부정확한 값을 넣으면 안된다는 리뷰엔 완전히 동의했어

사실 아직도 `cast_corpus_data`에서도  _"None으로 채우는 편이 훨 좋아보이는데"_ 라는 리뷰에는 동의가 잘 안 돼.
전부 None이어서 아예 모듈을 사용하지 못하는 것 보다는 이렇게라도 넣어준 뒤, passage filter와 함께 사용한다면 그 자체로 효과와 의미가 있을 거라고 생각하기 때문이야.

그치만, cast에서도 어짜피 last_modified_time에 `datetime.now()`가 특별히 제 기능을 못한다는 걸 생각했을 때, 지금부터라도 확실하지 않으면 None으로 표현하는 게 우리가 지게 될(?) 위험 부담을 줄여준다는 거엔 동의해. 
그래서 일단 리뷰 받은대로 수정해놓긴 했어.

vkehfdl1 replied:
난 여기서 prev & next에 '잘못된 가능성'이 있는 값을 넣는것과, `last_modified_datetime`에 `datetime.now()`를 넣는 것은 명확한 차이점이 존재한다고 생각. 

### datetime의 경우
1. 만약 `last_modified_datetime`이 주어져 있지 않은 경우, corpus_df를 만드는 시각을 포함시킴. 만약 이렇게만 한다면 *당연히* 별 의미가 없음. 
2. 그러나, 해당 corpus_df에 추후에 뭔가를 추가할 때 큰 의미가 있어짐. 예를 들어 보자.
내가 법률 문서를 corpus_df로 한달 전에 만들었다고 하자. 그런데 한 달 동안 '유사한' 내용이지만 법령이 개정된거임. 그리고 개정된 법령을 corpus에 추가해줌. 이런 경우, 한달 전 법령은 한달 전 datetime을, 지금 추가한 것은 현재 datetime을 가지게 되면서 'time reranker'나 'recency filter'에 큰 의미가 생김.
3. 위와 같은 이유로 datetime이 없을 경우 일괄 `datetime.now()`를 넣는 것에, None으로 넣는 것보다 더 의미가 있음. 지금 쓰는 것보다는, 나중을 위한 기록적 의미로.


### prev_id 및 next_id의 경우
1. 한 번 잘못된 값을 넣어 놓으면 계속해서 잘못된 값으로 유지 될 것임. 이는 `datetime`의 경우와 같은 이익이 없음.
2. 게다가 `datetime.now()`와 같이 나름 의미가 있는 것도 아니고, **완전히** 잘못된 값이 들어갈 수 있다는 위험이 있음.
3. prev 및 next를 못 쓸 때는, 깔끔하게 사용 가능 모듈에서 제외하는 편이 좋다고 봄. 그래야 나중에 '제대로 prev와 next' 값이 갖춰진 corpus를 추가했을 때 제대로 모듈을 사용할 수 있을 것임. 
4. 정리하자면 'corpus의 정확도 > 사용 가능한 모듈의 수'라고 생각함. 거짓 데이터는 최악이라고 생각.

충분한 설명이 되었을까?

bwook00 replied:
완전히 설득되었어. 마지막 수정해둔대로 쭉 진행하면 될듯 👍

vkehfdl1 reviewed (approved):
LGTM finally 👍
